util/src/main/java/com/google/cloud/hadoop/util/GrpcErrorTypeExtractor.java

```diff
@@ -34,8 +39,40 @@ public ErrorType getErrorType(Exception error) {
         return ErrorType.NOT_FOUND;
       case OUT_OF_RANGE:
         return ErrorType.OUT_OF_RANGE;
+      case ALREADY_EXISTS:
+        return ErrorType.ALREADY_EXISTS;
+      case FAILED_PRECONDITION:
+        return ErrorType.FAILED_PRECONDITION;
       default:
         return ErrorType.UNKNOWN;
     }
   }
+
+  @Override
+  public boolean bucketAlreadyExists(Exception e) {
+    ErrorType errorType = getErrorType(e);
+    if (errorType == ErrorType.ALREADY_EXISTS) return true;
+    // The gRPC API currently throws a FAILED_PRECONDITION status code instead of ALREADY_EXISTS,
+    // so we handle both these conditions in the interim. This can be removed once the status codes
+    // are fixed.
+    else if (errorType == ErrorType.FAILED_PRECONDITION) {
+      StatusRuntimeException statusRuntimeException = getStatusRuntimeException(e);
+      return statusRuntimeException != null
+          && BUCKET_ALREADY_EXISTS_MESSAGE.equals(statusRuntimeException.getMessage());
```

Comment:
Is there an error code? Checking for error message is not fool proof. Can it return different error message based on caller locale?

Reply:
We are checking the status code on line 58, but it is a generic FAILED_PRECONDITION code so added another level of check via error message. I dont think it returns different message based on caller locale as we have similar message check for json api as well - https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/fabf01d601537f9f663ff700e457809edbd172fe/util/src/main/java/com/google/cloud/hadoop/util/ApiErrorExtractor.java#L224